File: ios/App/GolfWatch Watch App/Practice/SwingDetector.swift
import Foundation
import simd

/// Phases of a golf swing, detected from wrist motion only.
enum SwingPhase {
    case idle
    case backswing
    case topOfBackswing
    case downswing
    case impact
    case finished
}

/// State machine that walks wrist-motion samples through the swing phases and
/// reports when a full swing has completed. Thresholds are empirical and will
/// need on-watch tuning — they're surfaced here (not buried) for exactly that
/// reason.
///
/// Detection is intentionally simple (no ML): it keys off angular-speed onset,
/// a dominant-axis sign reversal at the top, the acceleration spike near
/// impact, and a settle window for the finish.
final class SwingDetector {
    private(set) var phase: SwingPhase = .idle
    private(set) var tStart: TimeInterval?
    private(set) var tTop: TimeInterval?
    private(set) var tImpact: TimeInterval?
    private(set) var tFinish: TimeInterval?
    /// True when the downswing decayed with no impact spike — a rehearsal /
    /// air swing. The calculator labels these and the phone keeps them out of
    /// the real-swing stats.
    private(set) var isAirSwing = false
    /// Peak linear-accel magnitude (g) seen from the downswing through impact.
    /// A real ball strike spikes far harder than a turf brush / rehearsal, so
    /// round-mode uses this to reject practice swings that graze the ground.
    private(set) var peakImpactG = 0.0

    // --- tunable thresholds ---
    /// Below this angular speed (rad/s) the wrist is considered quiet.
    private let quietOmega = 0.6
    /// Angular speed (rad/s) that marks the start of an intentional move.
    private let startOmega = 2.5
    /// Linear-accel spike (g) that marks the impact window.
    private let impactAccel = 2.5
    /// How long (s) motion must stay quiet after impact to call the finish.
    private let finishSettleS = 0.4
    /// If no impact spike lands within this long after the downswing starts,
    /// treat the swing as an air swing (impact ≈ the peak-speed moment).
    private let airSwingTimeoutS = 0.8

    // --- stuck-phase timeouts ---
    //
    // Without these the machine has NO route back to idle except completing a
    // swing, and that loses real shots on the course. Every non-idle phase is
    // entered on motion that only MIGHT be a swing: pulling a club from the
    // bag, a waggle, an arm swinging while walking. If that motion doesn't go
    // on to satisfy the next transition, the phase simply persists — and the
    // golfer's next real swing is then read as the CONTINUATION of the stale
    // one rather than a new swing from idle. Its backswing gets consumed as a
    // downswing, the window is wrong, and the shot is either mis-measured or
    // dropped as an air swing.
    //
    // Generous next to a real swing (takeaway to finish is ~1.5-2.5s), so a
    // slow deliberate backswing or a pause at the top is never cut short. These
    // only fire on motion that was never a swing to begin with.
    private let backswingTimeoutS = 3.0
    private let topOfBackswingTimeoutS = 3.0
    /// A settle that never arrives — the arm kept moving after a strike (walking
    /// off, or straight into a re-tee). The strike itself was real, so this
    /// CLOSES the swing rather than discarding it.
    private let impactTimeoutS = 4.0

    /// When the current phase was entered. Drives the timeouts above.
    private var phaseSince: TimeInterval?

    private var dominantAxis = 0
    private var lastSignAtDominant: Double = 0
    private var quietSince: TimeInterval?
    private var downswingStart: TimeInterval?
    private var peakOmega = 0.0
    private var peakOmegaTime: TimeInterval?

    /// Feed one sample. Returns `.finished` exactly once, on the sample that
    /// completes a swing; otherwise nil.
    func advance(with s: MotionSample) -> SwingPhase? {
        // Deal with a phase that is going nowhere BEFORE the transitions below,
        // so a stale phase never gets to consume this sample as if it were part
        // of its own swing.
        if phase != .idle {
            switch resolveStalledPhase(at: s.t) {
            case .none: break
            case .abandoned: return nil
            case .completed: return .finished
            }
        }

        switch phase {
        case .idle:
            if s.angularSpeed > startOmega {
                enter(.backswing, at: s.t)
                tStart = s.t
                dominantAxis = indexOfMaxAbs(s.rotationRate)
                lastSignAtDominant = signOf(s.rotationRate[dominantAxis])
            }

        case .backswing:
            let sg = signOf(s.rotationRate[dominantAxis])
            // The dominant rotation axis reverses direction at a low-speed
            // pivot = top of the backswing.
            if sg != 0 && sg != lastSignAtDominant && s.angularSpeed < startOmega {
                enter(.topOfBackswing, at: s.t)
                tTop = s.t
            }
            if sg != 0 { lastSignAtDominant = sg }

        case .topOfBackswing:
            if s.angularSpeed > startOmega {
                enter(.downswing, at: s.t)
                downswingStart = s.t
                peakOmega = s.angularSpeed
                peakOmegaTime = s.t
            }

        case .downswing:
            if s.angularSpeed > peakOmega {
                peakOmega = s.angularSpeed
                peakOmegaTime = s.t
            }
            if s.linearAccelMag > peakImpactG { peakImpactG = s.linearAccelMag }
            if s.linearAccelMag >= impactAccel {
                // Real strike.
                enter(.impact, at: s.t)
                tImpact = s.t
                quietSince = nil
            } else if let ds = downswingStart, s.t - ds > airSwingTimeoutS {
                // No contact within the window → rehearsal / air swing. Use the
                // peak-speed moment as the impact reference so the timing
                // metrics still work.
                enter(.impact, at: s.t)
                isAirSwing = true
                tImpact = peakOmegaTime ?? s.t
                quietSince = nil
            }

        case .impact:
            if s.linearAccelMag > peakImpactG { peakImpactG = s.linearAccelMag }
            if s.angularSpeed < quietOmega {
                if let since = quietSince {
                    if s.t - since > finishSettleS {
                        enter(.finished, at: s.t)
                        tFinish = s.t
                        return .finished
                    }
                } else {
                    quietSince = s.t
                }
            } else {
                quietSince = nil
            }

        case .finished:
            break
        }
        return nil
    }

    /// Slice the completed swing out of the rolling buffer.
    func takeCompletedWindow(from buffer: [MotionSample]) -> SwingWindow? {
        guard let a = tStart, let top = tTop, let imp = tImpact, let f = tFinish else { return nil }
        let samples = buffer.filter { $0.t >= a && $0.t <= f }
        guard samples.count > 3 else { return nil }
        return SwingWindow(
            samples: samples, tStart: a, tTop: top, tImpact: imp, tFinish: f,
            isAirSwing: isAirSwing, peakImpactG: peakImpactG
        )
    }

    /// Move to `next`, remembering when — the timeouts are measured from here.
    private func enter(_ next: SwingPhase, at t: TimeInterval) {
        phase = next
        phaseSince = t
    }

    private enum StallOutcome {
        /// Still within its allowance — carry on.
        case none
        /// Nothing had been struck yet, so the phase was simply dropped.
        case abandoned
        /// A real strike was already detected; the swing was closed out.
        case completed
    }

    /// Handle a phase that has outstayed its allowance.
    ///
    /// Only HARD timeouts, deliberately. An earlier version also abandoned a
    /// phase whose motion had gone quiet, which reads as the obvious signal —
    /// but at the top of the backswing the wrist IS quiet, so it fired during
    /// real swings. Resetting there is worse than the stall it cured: the
    /// downswing then re-enters `.backswing` from idle and the impact spike
    /// arrives in a phase that cannot recognise it, losing the shot outright.
    ///
    /// The allowances below are long next to a real swing, so they only ever
    /// fire on motion that was never a swing. Recovering in three seconds
    /// instead of never is what matters here — shots are minutes apart.
    private func resolveStalledPhase(at t: TimeInterval) -> StallOutcome {
        guard let since = phaseSince else {
            // No entry stamp — adopt this sample so the timeout has an origin.
            phaseSince = t
            return .none
        }
        // Clock went backwards (CoreMotion re-bases timestamps across a
        // restart). Re-anchor rather than treating every sample as an eternity.
        if t < since {
            phaseSince = t
            return .none
        }

        switch phase {
        case .backswing where t - since > backswingTimeoutS,
             .topOfBackswing where t - since > topOfBackswingTimeoutS:
            // Nothing was struck, so there is nothing to lose by dropping it.
            reset()
            return .abandoned

        case .impact where t - since > impactTimeoutS:
            // A real impact spike WAS seen; only the settle never came. Closing
            // the swing here records the shot instead of throwing it away.
            tFinish = t
            phase = .finished
            return .completed

        default:
            // `.downswing` has its own escape (airSwingTimeoutS) and `.finished`
            // is consumed by the caller on the next line.
            return .none
        }
    }

    func reset() {
        phase = .idle
        phaseSince = nil
        tStart = nil; tTop = nil; tImpact = nil; tFinish = nil
        isAirSwing = false
        peakImpactG = 0
        quietSince = nil
        dominantAxis = 0
        lastSignAtDominant = 0
        downswingStart = nil
        peakOmega = 0
        peakOmegaTime = nil
    }

    // --- helpers ---
    private func signOf(_ v: Double) -> Double {
        if v > 0 { return 1 }
        if v < 0 { return -1 }
        return 0
    }

    private func indexOfMaxAbs(_ v: simd_double3) -> Int {
        let ax = abs(v.x), ay = abs(v.y), az = abs(v.z)
        if ax >= ay && ax >= az { return 0 }
        if ay >= ax && ay >= az { return 1 }
        return 2
    }
}

File: android/wear/src/main/java/com/golfroundtracker/app/wear/SwingDetector.kt
package com.golfroundtracker.app.wear

import kotlin.math.abs

enum class SwingPhase { IDLE, BACKSWING, TOP_OF_BACKSWING, DOWNSWING, IMPACT, FINISHED }

/**
 * State machine walking wrist motion through the swing phases.
 *
 * A FAITHFUL PORT of `ios/App/GolfWatch Watch App/Practice/SwingDetector.swift`
 * — same phases, same thresholds, same stall handling — because that logic was
 * tuned against real swings on real wrists and re-deriving it from first
 * principles on a second platform would throw that away. Keep the two in step;
 * where they must diverge, say so here rather than quietly drifting.
 *
 * Detection is deliberately not ML: angular-speed onset, a dominant-axis sign
 * reversal at the top, an acceleration spike at impact, and a settle window for
 * the finish.
 *
 * THE ONE THING THAT MAY NEED RE-TUNING IS `impactAccel`. Impact is a very short
 * spike, and whether it is caught at all depends on the sample rate the watch
 * actually delivers — CoreMotion guarantees 100 Hz (200 on newer hardware),
 * Android guarantees nothing. `SwingMotionService` measures the real rate for
 * exactly this reason.
 */
class SwingDetector {

    var phase: SwingPhase = SwingPhase.IDLE
        private set

    var tStart: Double? = null; private set
    var tTop: Double? = null; private set
    var tImpact: Double? = null; private set
    var tFinish: Double? = null; private set

    /** Downswing decayed with no impact spike — a rehearsal. */
    var isAirSwing = false; private set

    /** Peak linear-accel (g) from the downswing through impact. A ball strike
     *  spikes far harder than a turf brush, which is how round mode rejects
     *  practice swings. */
    var peakImpactG = 0.0; private set

    // --- tunable thresholds (values from the Swift original) ---
    private val quietOmega = 0.6          // rad/s below which the wrist is quiet
    private val startOmega = 2.5          // rad/s marking an intentional move
    private val impactAccel = 2.5         // g spike marking the impact window
    private val finishSettleS = 0.4       // s of quiet before calling the finish
    private val airSwingTimeoutS = 0.8    // s after downswing with no contact

    // --- stuck-phase timeouts ---
    //
    // Without these there is no route back to IDLE except completing a swing,
    // and that LOSES REAL SHOTS. Every non-idle phase is entered on motion that
    // only might be a swing — pulling a club from the bag, a waggle, an arm
    // swinging while walking. If that motion does not go on to satisfy the next
    // transition the phase persists, and the golfer's next real swing is read as
    // a continuation of the stale one: its backswing is consumed as a downswing,
    // the window is wrong, and the shot is mis-measured or dropped as air.
    //
    // Generous next to a real swing (takeaway to finish is 1.5-2.5 s), so a slow
    // backswing or a pause at the top is never cut short.
    private val backswingTimeoutS = 3.0
    private val topOfBackswingTimeoutS = 3.0
    /** A settle that never arrives — the arm kept moving after a real strike.
     *  CLOSES the swing rather than discarding it. */
    private val impactTimeoutS = 4.0

    private var phaseSince: Double? = null
    private var dominantAxis = 0
    private var lastSignAtDominant = 0.0
    private var quietSince: Double? = null
    private var downswingStart: Double? = null
    private var peakOmega = 0.0
    private var peakOmegaTime: Double? = null

    /** Feed one sample. Returns FINISHED exactly once, on the completing sample. */
    fun advance(s: MotionSample): SwingPhase? {
        // Resolve a phase that is going nowhere BEFORE the transitions below, so
        // a stale phase never consumes this sample as part of its own swing.
        if (phase != SwingPhase.IDLE) {
            when (resolveStalledPhase(s.t)) {
                StallOutcome.NONE -> Unit
                StallOutcome.ABANDONED -> return null
                StallOutcome.COMPLETED -> return SwingPhase.FINISHED
            }
        }

        when (phase) {
            SwingPhase.IDLE -> {
                if (s.angularSpeed > startOmega) {
                    enter(SwingPhase.BACKSWING, s.t)
                    tStart = s.t
                    dominantAxis = indexOfMaxAbs(s)
                    lastSignAtDominant = signOf(s.rotationAt(dominantAxis))
                }
            }

            SwingPhase.BACKSWING -> {
                val sg = signOf(s.rotationAt(dominantAxis))
                // The dominant rotation axis reverses at a low-speed pivot —
                // that is the top of the backswing.
                if (sg != 0.0 && sg != lastSignAtDominant && s.angularSpeed < startOmega) {
                    enter(SwingPhase.TOP_OF_BACKSWING, s.t)
                    tTop = s.t
                }
                if (sg != 0.0) lastSignAtDominant = sg
            }

            SwingPhase.TOP_OF_BACKSWING -> {
                if (s.angularSpeed > startOmega) {
                    enter(SwingPhase.DOWNSWING, s.t)
                    downswingStart = s.t
                    peakOmega = s.angularSpeed
                    peakOmegaTime = s.t
                }
            }

            SwingPhase.DOWNSWING -> {
                if (s.angularSpeed > peakOmega) {
                    peakOmega = s.angularSpeed
                    peakOmegaTime = s.t
                }
                if (s.linearAccelMag > peakImpactG) peakImpactG = s.linearAccelMag
                val ds = downswingStart
                if (s.linearAccelMag >= impactAccel) {
                    enter(SwingPhase.IMPACT, s.t)
                    tImpact = s.t
                    quietSince = null
                } else if (ds != null && s.t - ds > airSwingTimeoutS) {
                    // No contact in the window → rehearsal. Use the peak-speed
                    // moment as the impact reference so timing metrics still work.
                    enter(SwingPhase.IMPACT, s.t)
                    isAirSwing = true
                    tImpact = peakOmegaTime ?: s.t
                    quietSince = null
                }
            }

            SwingPhase.IMPACT -> {
                if (s.linearAccelMag > peakImpactG) peakImpactG = s.linearAccelMag
                if (s.angularSpeed < quietOmega) {
                    val since = quietSince
                    if (since == null) {
                        quietSince = s.t
                    } else if (s.t - since > finishSettleS) {
                        enter(SwingPhase.FINISHED, s.t)
                        tFinish = s.t
                        return SwingPhase.FINISHED
                    }
                } else {
                    quietSince = null
                }
            }

            SwingPhase.FINISHED -> Unit
        }
        return null
    }

    /** Slice the completed swing out of the rolling buffer. */
    fun takeCompletedWindow(buffer: List<MotionSample>): SwingWindow? {
        val a = tStart ?: return null
        val top = tTop ?: return null
        val imp = tImpact ?: return null
        val f = tFinish ?: return null
        val samples = buffer.filter { it.t in a..f }
        if (samples.size <= 3) return null
        return SwingWindow(samples, a, top, imp, f, isAirSwing, peakImpactG)
    }

    private fun enter(next: SwingPhase, t: Double) {
        phase = next
        phaseSince = t
    }

    private enum class StallOutcome { NONE, ABANDONED, COMPLETED }

    /**
     * Handle a phase that has outstayed its allowance.
     *
     * ONLY HARD TIMEOUTS, deliberately. An earlier iOS version also abandoned a
     * phase whose motion had gone quiet — the obvious signal — but at the top of
     * the backswing the wrist IS quiet, so it fired during real swings. The cure
     * was worse than the stall: the downswing then re-entered BACKSWING from
     * idle and the impact spike arrived in a phase that could not recognise it,
     * losing the shot outright.
     */
    private fun resolveStalledPhase(t: Double): StallOutcome {
        val since = phaseSince
        if (since == null) {
            phaseSince = t
            return StallOutcome.NONE
        }
        // Clock went backwards. On Android this is rarer than on iOS (the sensor
        // clock is monotonic since boot) but a re-registration can still re-base
        // it, and treating that as an eternity would abandon a live swing.
        if (t < since) {
            phaseSince = t
            return StallOutcome.NONE
        }

        return when {
            phase == SwingPhase.BACKSWING && t - since > backswingTimeoutS ||
                phase == SwingPhase.TOP_OF_BACKSWING && t - since > topOfBackswingTimeoutS -> {
                // Nothing struck, so nothing to lose by dropping it.
                reset()
                StallOutcome.ABANDONED
            }

            phase == SwingPhase.IMPACT && t - since > impactTimeoutS -> {
                // A real spike WAS seen; only the settle never came. Closing the
                // swing here records the shot instead of throwing it away.
                tFinish = t
                phase = SwingPhase.FINISHED
                StallOutcome.COMPLETED
            }

            // DOWNSWING has its own escape (airSwingTimeoutS); FINISHED is
            // consumed by the caller.
            else -> StallOutcome.NONE
        }
    }

    fun reset() {
        phase = SwingPhase.IDLE
        phaseSince = null
        tStart = null; tTop = null; tImpact = null; tFinish = null
        isAirSwing = false
        peakImpactG = 0.0
        quietSince = null
        dominantAxis = 0
        lastSignAtDominant = 0.0
        downswingStart = null
        peakOmega = 0.0
        peakOmegaTime = null
    }

    private fun signOf(v: Double): Double = when {
        v > 0 -> 1.0
        v < 0 -> -1.0
        else -> 0.0
    }

    private fun indexOfMaxAbs(s: MotionSample): Int {
        val ax = abs(s.gx); val ay = abs(s.gy); val az = abs(s.gz)
        return when {
            ax >= ay && ax >= az -> 0
            ay >= ax && ay >= az -> 1
            else -> 2
        }
    }
}
